pkg/kubelet/metrics/collectors/cri_metrics.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collectors
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/component-base/metrics"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+	"k8s.io/klog/v2"
+)
+
+type criMetricsCollector struct {
+	metrics.BaseStableCollector
+	descriptors           map[string]*metrics.Desc
+	listPodSandboxMetrics func(context.Context) ([]*runtimeapi.PodSandboxMetrics, error)
+}
+
+// Check if criMetricsCollector implements necessary interface
+var _ metrics.StableCollector = &criMetricsCollector{}
+
+// NewCRIMetricsCollector implements the metrics.Collector interface
+func NewCRIMetricsCollector(ctx context.Context, listPodSandboxMetrics func(context.Context) ([]*runtimeapi.PodSandboxMetrics, error), listMetricDescriptors func(context.Context) ([]*runtimeapi.MetricDescriptor, error)) metrics.StableCollector {
+	descs, err := listMetricDescriptors(ctx)
+	if err != nil {
```

Comment:
Does this mean that if we fail the initial descriptor list, we never get any help text?

Reply:
yeah. I suppose we could re-request if we see the structure is empty, though I think we'd also have to add some locking to do so.

Reply:
thinking about it more, I think for now this is sufficient. Since they're separated calls, we can always add additional calls in the future, but this is conceptually simple and avoids excessive work when CRI doesn't support the call